### PR TITLE
Add ConformerEncoderV2

### PR DESCRIPTION
**Pull request by Judyxujj**
For the second version of Conformer, two additional options are added:
1. enable swapping MHSA and Conv module
2. enable adding auxiliary losses on intermediate layers

### CODE REVIEW

File: i6_models/assemblies/conformer/conformer_v2.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        mhsa_cfg: Configuration for ConformerMHSAV1
+        conv_cfg: Configuration for ConformerConvolutionV1
+        swap_mhsa_conv: swap the execution order of MHSA and Conv module
+    """
+
+    # nested configurations
+    ff_cfg: ConformerPositionwiseFeedForwardV1Config
+    mhsa_cfg: ConformerMHSAV1Config
+    conv_cfg: ConformerConvolutionV1Config
+    swap_mhsa_conv: bool = True
```
Comment (mmz33) [inline i6_models/assemblies/conformer/conformer_v2.py:35]: Probably better to have this `False` by default here.

File: i6_models/assemblies/conformer/conformer_v2.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+    def __init__(self, cfg: ConformerEncoderV2Config):
+        """
+        :param cfg: conformer encoder configuration with subunits for frontend and conformer blocks
+        """
+        super().__init__()
+
+        self.frontend = cfg.frontend()
+        self.module_list = torch.nn.ModuleList([ConformerBlockV2(cfg.block_cfg) for _ in range(cfg.num_layers)])
+        self.output_layers = list(cfg.loss_scales.keys())
```
Comment (mmz33) [inline i6_models/assemblies/conformer/conformer_v2.py:108]: I am wondering if you really need to pass the whole `loss_scales` dict since the scales are not used in the `forward function`. Instead, you can pass the list of indicies directly.

Comment (Judyxujj) [reply i6_models/assemblies/conformer/conformer_v2.py:108]: but later in train step, we need the loss scale, I think it would be easier to get the scales from `ctc.conformer. loss_scales` instead of passing it as an argument to train step function

File: i6_models/assemblies/conformer/conformer_v2.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        assert cfg.num_layers in self.output_layers, "The final layer must be included in loss layers"
+
+    def forward(
+        self, data_tensor: torch.Tensor, sequence_mask: torch.Tensor
```
Comment (albertz) [inline i6_models/assemblies/conformer/conformer_v2.py:112]:
```suggestion
        self, data_tensor: torch.Tensor, /, sequence_mask: torch.Tensor
```
?
(Would be consistent to above.)

File: i6_models/assemblies/conformer/conformer_v2.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    ) -> Tuple[List[torch.Tensor], torch.Tensor]:
+        """
+        :param data_tensor: input tensor of shape [B, T', F]
+        :param sequence_mask: mask tensor where 1 defines positions within the sequence and 0 outside, shape: [B, T']
+        :return: (output, out_seq_mask)
+            where output is torch.Tensor of shape [B, T, F'],
```
Comment (albertz) [inline i6_models/assemblies/conformer/conformer_v2.py:118]:
```suggestion
        :return: (outputs, out_seq_mask)
            where outputs is a list of torch.Tensor of shape [B, T, F']
            for each of the layers in output_layers,
```

File: i6_models/assemblies/conformer/conformer_v2.py (hunk 1 above)
Comment (JackTemaki) [reply i6_models/assemblies/conformer/conformer_v2.py:35]: After the latest experiments I would also agree to have it false, we still do not see consistent and reliable improvements...

File: i6_models/assemblies/conformer/conformer_v2.py (hunk 2 above)
Comment (JackTemaki) [reply i6_models/assemblies/conformer/conformer_v2.py:108]: I think Mohammad is right here, the `loss_scales` should be attached to your (own. custom) model config, not the encoder config. You can even put the parameter for which layers to return into the forward instead of the class That way a user can have different auxiliary losses during different training stages without having to modify the encoder code here.

File: i6_models/assemblies/conformer/conformer_v2.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    def __init__(self, cfg: ConformerEncoderV2Config):
+        """
+        :param cfg: conformer encoder configuration with subunits for frontend and conformer blocks
+        """
+        super().__init__()
+
+        self.frontend = cfg.frontend()
+        self.module_list = torch.nn.ModuleList([ConformerBlockV2(cfg.block_cfg) for _ in range(cfg.num_layers)])
+        self.output_layers = list(cfg.loss_scales.keys())
+        assert cfg.num_layers in self.output_layers, "The final layer must be included in loss layers"
```
Comment (JackTemaki) [inline i6_models/assemblies/conformer/conformer_v2.py:109]: when you put this into the forward instead of the class init, you could even say the user is allowed to not access the final layer, and then you do a `min(len(...), max(return_layers))` to determine how long the module loop should work.

**Review by JackTemaki — CHANGES_REQUESTED** [review]

File: i6_models/assemblies/conformer/conformer_v2.py (hunk 5 above)
Comment (Judyxujj) [reply i6_models/assemblies/conformer/conformer_v2.py:109]: I don't fully get the point here. If the final layer is not included and user is allowed to not access the final layer, then it would waste the memory to load the layers between [max(loss_layers),  num_layers].

Comment (JackTemaki) [reply i6_models/assemblies/conformer/conformer_v2.py:109]: Yes, it would waste the memory but not the computation. I am just saying this is an unnecessary limitation in my view. I would just let this be user choice and not enforce it.

File: i6_models/assemblies/conformer/conformer_v2.py (hunk 2 above)
Comment (JackTemaki) [reply i6_models/assemblies/conformer/conformer_v2.py]: This is not resolved, I would still vote to have a `return_layers: List[int]` in the forward instead of having this static in the config. @mmz33 @albertz what is your opinion on this?

Comment (JackTemaki) [reply i6_models/assemblies/conformer/conformer_v2.py]: @Judyxujj I know this might be overkill, but one use case would be to have "moving" auxiliary losses, that are not at fixed layers but start in early layers and move higher over the course of training.

File: i6_models/assemblies/conformer/conformer_v2.py (hunk 5 above)
Comment (Judyxujj) [reply i6_models/assemblies/conformer/conformer_v2.py:109]: could you explain more about `min(len(...), max(return_layers))`, I don't understand the idea here

Comment (JackTemaki) [reply i6_models/assemblies/conformer/conformer_v2.py:109]: What I meant:
```python
        outputs = []
        for i in range(min(len(self.module_list), max(return_layers)):
            x = self.module_list[i](x, sequence_mask)  # [B, T, F']
            if i + 1 in return_layers:
                outputs.append(x)
                ```
But probably much better idea:
```python
        outputs = []
        assert max(return_layers) <= len(self.module_list)
        for i in range(max(return_layers)):
            x = self.module_list[i](x, sequence_mask)  # [B, T, F']
            if i + 1 in return_layers:
                outputs.append(x)
                ```

File: .github/workflows/black.yml [hunk 6]
```diff
@@ -10,8 +10,8 @@ jobs:
   check-black-formatting:
     runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@v2
-    - uses: actions/setup-python@v2
+    - uses: actions/checkout@v4
+    - uses: actions/setup-python@v4
```
Comment (albertz) [inline .github/workflows/black.yml:14]: I think this does not belong to the PR here. Is this needed for the PR? I would do a separate PR for this, and then rebase this one here.

File: .github/workflows/model_tests.yml [hunk 7]
```diff
@@ -14,7 +14,7 @@ jobs:
         with:
           repository: "rwth-i6/i6_models"
           path: ""
-      - uses: actions/setup-python@v2
+      - uses: actions/setup-python@v4
```
Comment (albertz) [inline .github/workflows/model_tests.yml:17]: See above.

File: i6_models/assemblies/conformer/conformer_v2.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
+from __future__ import annotations
```
Comment (albertz) [inline i6_models/assemblies/conformer/conformer_v2.py:1]: At some point, maybe in the (missing) docstring of the module, you should very briefly explain what's new in `conformer_v2` (just bullet points).

File: i6_models/assemblies/conformer/conformer_v2.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+
+@dataclass
+class ConformerBlockV2Config(ModelConfiguration):
+    """
+    Attributes:
+        ff_cfg: Configuration for ConformerPositionwiseFeedForwardV1
+        mhsa_cfg: Configuration for ConformerMHSAV1
+        conv_cfg: Configuration for ConformerConvolutionV1
+        modules: List of modules to use for ConformerBlockV2,
+                       "ff" for feed forward module, "mhsa" for multi-head self attention module, "conv" for conv module
```
Comment (albertz) [inline i6_models/assemblies/conformer/conformer_v2.py:30]: Indentation looks off. I think it should be indented by 4 spaces.
```suggestion
        modules: List of modules to use for ConformerBlockV2,
            "ff" for feed forward module, "mhsa" for multi-head self attention module, "conv" for conv module
```

File: i6_models/assemblies/conformer/conformer_v2.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+        self.frontend = cfg.frontend()
+        self.module_list = torch.nn.ModuleList([ConformerBlockV2(cfg.block_cfg) for _ in range(cfg.num_layers)])
+
+    def forward(
+        self, data_tensor: torch.Tensor, /, sequence_mask: torch.Tensor, return_layers: Optional[List[int]] = None
+    ) -> Tuple[List[torch.Tensor], torch.Tensor]:
+        """
+        :param data_tensor: input tensor of shape [B, T', F]
+        :param sequence_mask: mask tensor where 1 defines positions within the sequence and 0 outside, shape: [B, T']
+        :param return_layers: list of layer indices specifying which layers to return, starting from 1
```
Comment (albertz) [inline i6_models/assemblies/conformer/conformer_v2.py:130]: Why do you start counting from 1? To me (as someone who writes code) this was unexpected. I'm usually expecting always that indices start from 0 (at least in code).

File: .github/workflows/black.yml (hunk 6 above)
Comment (JackTemaki) [reply .github/workflows/black.yml:14]: https://github.com/rwth-i6/i6_models/pull/52

File: i6_models/assemblies/conformer/conformer_v2.py (hunk 8 above)
Comment (Judyxujj) [reply i6_models/assemblies/conformer/conformer_v2.py:1]: they are now added to the doctoring of the module

File: i6_models/assemblies/conformer/conformer_v2.py (hunk 10 above)
Comment (Judyxujj) [reply i6_models/assemblies/conformer/conformer_v2.py]: The counting is changed to start from 0

**Review by JackTemaki — APPROVED** [review]
Did a full setup run and it works as intended

**Review by Atticus1806 — APPROVED** [review]

**Review by SimBe195 — APPROVED** [review]
